src/routes/sessions.js

```diff
@@ -62,50 +60,76 @@ module.exports = function (app, opts, dependencies) {
     };
   }
 
+  function getTwoFactorResponse(user) {
+    if (process.env.LIANA_2FA_SECRET  === undefined) {
+      logger.error('Cannot use the two factor authentication because the environment variable "LIANA_2FA_SECRET" is not set.');
+      throw new Error();
+    }
+
+    if (process.env.LIANA_2FA_SECRET.length !== 20) {
+      logger.error('The environment variable "LIANA_2FA_SECRET" must be 20 characters long.');
+      logger.error('You can generate it using this command: `$ openssl rand -hex 10`');
+      throw new Error();
+    }
+
+    if (user.two_factor_authentication_configured) {
+      return { twoFactorAuthenticationEnabled: true };
+    } else {
+      const partialUserSecret = user.partial_user_secret;
+      const userSecret = new UserSecretCreator(partialUserSecret, process.env.LIANA_2FA_SECRET)
+        .perform();
+
+      return {
+        twoFactorAuthenticationEnabled: true,
+        userSecret,
+      };
+    }
+  }
+
   function loginWithPassword(request, response) {
     const renderingId = request.body.renderingId;
     const envSecret = opts.envSecret;
+    const { email, password } = request.body;
 
-    return ipWhitelist.retrieve(envSecret)
-      .then(function () { return new AllowedUsersFinder(renderingId, envSecret).perform(); })
-      .then(function (allowedUsers) {
-        if (allowedUsers.length === 0) {
-          throw new Error(errorMessages.SESSION.NO_USERS);
-        }
+    (async function asyncWrapper() {
+      await ipWhitelist.retrieve(envSecret);
+      const user = await new AuthorizationFinder(renderingId, email, password, envSecret)
+        .perform();
 
-        const user = _.find(allowedUsers, function (allowedUser) {
-          return allowedUser.email === request.body.email;
-        });
+      if (!user) {
+        throw new Error();
+      }
 
-        if (user === undefined) {
-          throw new Error();
-        }
+      if (user.two_factor_authentication_enabled) {
+        const responseData = getTwoFactorResponse(user);
+        return response.send(responseData);
+      }
 
-        return bcrypt.compare(request.body.password, user.password)
-          .then(function (isEqual) {
-            if (!isEqual) { throw new Error(); }
-            return user;
-          });
-      })
-      .then(sendToken(response, renderingId))
-      .catch(formatAndSendError(response));
+      await sendToken(response, renderingId)(user);
```

Comment:
Maybe you can simplify the `sendToken` implementation without having to create a function in the function body.
(`sendToken(response, renderingId, user)`)